### ISSUE TITLE
support .exr and .webp format in image io

### ISSUE THREAD

**Issue by ghost** [kornia/kornia-rs#607]
## 🚀 Feature
support .exr and .webp format in image io

## Motivation

exr is the de-fatcto standard for HDR images in grahics softwares, and webp is widely used on the internet

## Additional context

______________________________________________________________________

#### Consider also to contribute to Kornia universe projects :)

<sub>

- [**Tutorials**](https://github.com/kornia/tutorials): our repository containing the tutorials.

</sub>

**Comment by edgarriba** [comment]
@nick008a this has to be exposed from [`kornia-rs`](https://github.com/kornia/kornia-rs) which internally uses the Rust `image` crate that seems to support those formats: https://crates.io/crates/image

Would you be open to help on that ?

**Comment by ghost** [comment]
I really hope I cloud, but I know nothing about rust...

also, is it possible to read from a bytearray like cv2.imdecode?

**Comment by edgarriba** [comment]
> also, is it possible to read from a bytearray like cv2.imdecode?

yes, but not exposed to kornia.io https://github.com/kornia/kornia-rs/blob/main/test/test_io.py#L55C34-L55C47

we could have a `decode_image` method for that

**Comment by AdarshGuptaa** [comment]
### Proposed Solution
I plan to implement _WebP_ support mirroring the existing _jpeg.rs_ and _png.rs_ patterns to ensure consistency with the functioning elements.

This will involve:

-     Creating a new module crates/kornia-io/src/webp.rs.
-     Updating the functional API in functional.rs to auto-detect .webp extensions.
-     Handling new error variants for WebP decoding/encoding as IOError::WebpDecodingError and IOError::WebpEncodingError


### API Design
Mirroring _jpeg.rs_ and _png.rs_ functions
```
// Writing (Encoding)
pub fn write_image_webp_rgb8<A: ImageAllocator>(
    file_path: impl AsRef<Path>,
    image: &Image<u8, 3, A>,
    quality: u8
) -> Result<(), IoError>;

pub fn write_image_webp_gray8<A: ImageAllocator>(
    file_path: impl AsRef<Path>,
    image: &Image<u8, 1, A>,
    quality: u8
) -> Result<(), IoError>;

// Reading (Decoding)
pub fn read_image_webp_rgb8(file_path: impl AsRef<Path>) -> Result<Rgb8<CpuAllocator>, IoError>;
pub fn read_image_webp_gray8(file_path: impl AsRef<Path>) -> Result<Gray8<CpuAllocator>, IoError>;

pub fn decode_image_webp_rgb8<A: ImageAllocator>(src: &[u8], dst: &mut Image<u8, 3, A>) -> Result<(), IoError>;
pub fn decode_image_webp_layout(src: &[u8]) -> Result<ImageLayout, IoError>;
```

### Dependencies

- image crate (already being utilised)

I have read through a fair bit of _crates/kornia-io_ and if my proposed solution is satisfactory then could I please be assigned this issue? If not, I am happy to take any recommendation for a change in approach.

**Comment by edgarriba** [comment]
which decoder are gonna use ?

**Comment by AdarshGuptaa** [comment]
I was a little complacent with my reasearch

I can convert the webp data into rgb8, rgba8 and gray8 format.

rgb to rgb8 and gray8
rgba to rgba8
can be handled by image::codecs::webp::WebPDecoder in a single line

image::codecs::webp::WebPDecoder has no support for 3 to 4 and 4 to 3 byte streams unlike jpeg_decoder
rgba to rgb8 and gray8 needs to be sliced for the 4th byte
rgb to rgba8 needs another additional step to add a byte

**Comment by edgarriba** [comment]
we should find an only decoder library, we removed in the past already image. Please, provide a more human readible response written by yourself, this seems the output of a bot

**Comment by AdarshGuptaa** [comment]
I didn't know image had been removed. I did some more digging

I am thinking of using: [image-webp](https://crates.io/crates/image-webp/0.2.4)

- Rust only crate similar to png, zune-jpeg and jpeg-encoder (already being used)
- Has an encoder and decoder only for webp files like you asked
- It has only 2 dependencies which are already in cargo.lock and are small

Does this sound okay to you?

**Comment by edgarriba** [comment]
sounds good then
